### PR TITLE
[paddle.linalg.qr] Add the Qr Operator

### PR DESCRIPTION
**Pull request by aoyulong**
### PR types
New features

### PR changes
OPs

### Describe
Add paddle.linalg.qr() function, Qr Op and Qr GPU/CPU kernel. 
**Note: Qr Op of this pull request doesn't have the backward kernels for now.**
qr

### CODE REVIEW

**Review by fuyinno4 — DISMISSED** [review]

File: python/paddle/__init__.py [hunk 1]
```diff
@@ -504,6 +505,7 @@
            'sqrt',
            'cholesky',
            'matrix_power',
+           'qr',
```
Comment (dingjiaweiww) [inline python/paddle/__init__.py:508]: qr 的暴露路径是paddle.linalg.qr, 不用暴露在paddle.qr 下面

File: python/paddle/tensor/linalg.py [hunk 2]
```diff
@@ -1490,6 +1490,65 @@ def matrix_power(x, n, name=None):
     return out
 
 
+def qr(x, mode="reduced", name=None):
+    r"""
+    Computes the QR decomposition of one matrix or batches of matrice.
+    Args:
```
Comment (dingjiaweiww) [inline python/paddle/tensor/linalg.py:1601]: args 之前加一个空行

File: python/paddle/tensor/linalg.py [hunk 3]
```diff
@@ -1490,6 +1490,65 @@ def matrix_power(x, n, name=None):
...
+            which means Q's shape is `[..., M, M]` and R's shape is `[..., M, N]`.
+            If mode = "r", qr op will only return reduced R matrix, which means
+            R's shape is `[..., K, N]`.
+            
+    Returns:
+        If mode = "reduced" or mode = "complete", qr will return a two tensor-tuple, which represents Q and R 
+        If mode = "r", qr will return a tensor which represents R.
+        
+    Examples:            
+            import paddle
```
Comment (dingjiaweiww) [inline python/paddle/tensor/linalg.py:1623]: import paddle 之前需要加入：.. code-block:: python
参考文档：http://agroup.baidu.com/paddlepaddle/md/article/3088623

**Review by dingjiaweiww — APPROVED** [review]

File: python/paddle/__init__.py (hunk 1 above)
Comment (dingjiaweiww) [inline python/paddle/__init__.py]: 线性代数API 不需要在paddle.* 下面暴露

**Review by dingjiaweiww — DISMISSED** [review]

**Review by dingjiaweiww — DISMISSED** [review]

File: python/paddle/__init__.py (hunk 1 above)
Comment (aoyulong) [reply python/paddle/__init__.py]: Done.

File: python/paddle/tensor/linalg.py (hunk 3 above)
Comment (aoyulong) [reply python/paddle/tensor/linalg.py:1623]: Done.

File: python/paddle/tensor/linalg.py (hunk 2 above)
Comment (aoyulong) [reply python/paddle/tensor/linalg.py:1601]: Done.

File: python/paddle/__init__.py [hunk 4]
```diff
@@ -105,6 +105,7 @@
 from .tensor.linalg import slogdet  # noqa: F401
 from .tensor.linalg import multi_dot  # noqa: F401
 from .tensor.linalg import matrix_power  # noqa: F401
+from .tensor.linalg import qr  # noqa: F401
```
Comment (jeff41404) [inline python/paddle/__init__.py]: should not add this line, otherwise we have paddle.qr

**Review by Superjomn — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/qr_op.h [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    compute_q = true;
+    reduced = true;
+  } else if (mode == "complete") {
+    compute_q = true;
+    reduced = false;
+  } else if (mode == "r") {
+    compute_q = false;
+    reduced = true;
+  } else {
+    PADDLE_ENFORCE(
```
Comment (chenwhql) [inline paddle/fluid/operators/qr_op.h]: 仅CUDA Kernel内的PADDLE_ENFORCE不需要加报错类型，这里也是需要加类型的，而且这里我理解用PADDLE_THROW更合适一些

File: paddle/fluid/operators/qr_op.h [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      }
+    }
+  }
+};
+
+template <typename DeviceContext, typename T>
+class QrGradKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const {
+    PADDLE_ENFORCE(
```
Comment (chenwhql) [inline paddle/fluid/operators/qr_op.h]: 同上

**Review by zhiqiu — DISMISSED** [review]
LGTM for ShareDataWith

**Review by TCChenlong — DISMISSED** [review]

**Review by zhhsplendid — DISMISSED** [review]
Discussed that ShareDataWith should not be applied on output tensers, but we are chasing release/2.2 so I will approve this PR and we will change in future PRs

**Review by chenwhql — DISMISSED** [review]
临近打tag，先approve，希望后面追加PR规范报错检查的使用

**Review by jzhang533 — APPROVED** [review]
lgtm

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

**Pull request merged**